### ISSUE TITLE
Should not close servlet outputstream

### ISSUE THREAD

**Issue by dleangen** [osgi/osgi.enroute.bundles#15]
From [line 484 of osgi.enroute.web.server.provider.WebServer](https://github.com/osgi/osgi.enroute.bundles/blob/master/osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/WebServer.java#L484):

				out.flush();
				out.close();
				rsp.getOutputStream().flush();
				rsp.getOutputStream().close();

I believe that the servlet should not close the output stream, but should instead allow the servlet container to do so.

I have correlated this exception below with execution of the code above:

    java.lang.IllegalStateException: Committed
	at org.eclipse.jetty.server.Response.resetBuffer(Response.java:1243)
	at javax.servlet.ServletResponseWrapper.resetBuffer(ServletResponseWrapper.java:195)
	at org.apache.felix.http.base.internal.dispatch.ServletResponseWrapper.sendError(ServletResponseWrapper.java:67)
	at org.apache.felix.http.base.internal.dispatch.ServletResponseWrapper.sendError(ServletResponseWrapper.java:61)
	at org.apache.felix.http.base.internal.dispatch.Dispatcher.dispatch(Dispatcher.java:133)
	at org.apache.felix.http.base.internal.DispatcherServlet.service(DispatcherServlet.java:61)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
    2016-01-09 16:21:04.417:WARN:oejs.HttpChannel:qtp351877391-31: Could not send response error 500: java.lang.IllegalStateException: Committed

I lack expertise to state the above with authority, but I am pretty confident that this is the case.

**Comment by pkriens** [comment]
I am pretty sure the servlet may close the stream; one of the most complex part of filters etc. is that they need to proxy the streams. I normally see this stuff happening when the servlet tries to get the OutputStream 2x or it gets the Writer version after it already had the output stream.

Since this is not a problem that happens normally, under what circumstances does this happen? I.e. your code?

**Comment by dleangen** [comment]
Hi,

My code does this:

            final String path = constructDispatchPathFrom( request, asset );
            final RequestDispatcher dispatcher = request.getRequestDispatcher( path );
            dispatcher.forward( request, response );

The request gets forwarded to WebServer, and outputs nothing. From the logs [edited]:

    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/css/a.css, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/java/b.js, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/java/c.js, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/css/d.css, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/css/e.css, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/java/f.js, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/java/g.js, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/java/h.js, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/java/i.js, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/java/j.js, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/images/logo.png, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/images/twitter_button.jpg, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/images/afaq_9001_inter_vf-small.png, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/images/contact.jpg, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/images/logo-very_small.gif, queryString = null]
    New dispatcher with RequestInfo[servletPath =/osgi.enroute.simple.server, pathInfo = /website/static-resources/images/favicon.ico, queryString = null]
    2016-01-09 22:32:01.806:WARN:oejs.ServletHandler:qtp351877391-33: /static-resources/images/favicon.ico

The only pattern I could consistently find was that the "Committed" exception was always thrown when those lines from 484 were executed. The exception is never thrown when those lines were not executed.

**Comment by dleangen** [comment]
Hi again,

The problem does not occur with a Filter, it occurs during a forward. The forwarding servlet should, I believe, not have to do any special handing, so the code above (repasted here) should work fine:

        String path = constructPath( ... );
        RequestDispatcher dispatcher = request.getRequestDispatcher( path );
        dispatcher.forward( request, response );

The spec seems pretty clear that the container is responsible for closing (and flushing) a stream, but I agree, it does not seem very clear as to whether or not it is acceptable to have a servlet (i.e. not the container) *also* close the stream. Although it may not be incorrect, it appears that it is not necessary.

Here are some excerpts from the spec (v3.1):


In 1.3 ("An Example"): 

    Once the servlet has finished processing the request, the servlet container ensures that the 
    response is properly flushed, and returns control back to the host Web server.

In 5.6 ("Closure of Response Object"):

    When a response is closed, the container must immediately flush all remaining content in the 
    response buffer to the client. The following events indicate that the servlet has satisfied the request 
    and that the response object is to be closed:
    ■ The termination of the service method of the servlet.
    ■ The amount of content specified in the setContentLength or setContentLengthLong method of the response has been greater than zero and has been written to the response.
    ■ The sendError method is called.
    ■ The sendRedirect method is called.
    ■ The complete method on AsyncContext is called.

In 9.4 ("The Forward Method"):

    Before the forward method of the RequestDispatcher interface returns without exception, the 
    response content must be sent and committed, and closed by the servlet container, unless the 
    request was put into the asynchronous mode. If an error occurs in the target of the 
    RequestDispatcher.forward() the exception may be propagated back through all the calling filters 
    and servlets and eventually back to the container.

(Don't know what that means: "unless the request was put into the asynchronous mode".)

**Comment by dleangen** [comment]
Ok, you were right (of course). :-)

It was correct to say that I found the error above to occur when the following was executed:

            out.flush();
            out.close();
            rsp.getOutputStream().flush();
            rsp.getOutputStream().close();

However, I tried removing those lines, and the error still occurred, so I need to correct my statement to mention that the problem occurs when (1) a request gets *forwarded* to osgi.enroute.web.server.provider and (2) the following block of code is executed:

			if (rq.getMethod().equalsIgnoreCase("GET")) {

				rsp.setContentLengthLong(range.length());
				OutputStream out = rsp.getOutputStream();

				try (FileInputStream file = new FileInputStream(c.file);) {
					FileChannel from = file.getChannel();
					WritableByteChannel to = Channels.newChannel(out);
					range.copy(from, to);
					from.close();
					to.close();
				}

				out.flush();
				out.close();
				rsp.getOutputStream().flush();
				rsp.getOutputStream().close();
			}

The above block of code is located at [line 471 of osgi.enroute.web.server.provider.WebServer](https://github.com/osgi/osgi.enroute.bundles/blob/master/osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/WebServer.java#L471).

I can confirm that the above is true, but I am at a loss as to why this happens. I tried inserting a 				rsp.reset() to see if an IllegalArgumentException would be thrown, but no such problem. Also, I can load the resource directly (i.e. when there is not a forward) without a problem.

:-(
